File: tests/c/decoder_framed.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "serialization_test.sf.h"
#include "struct_frame_default_frame.h"

// Decoder for cross-platform testing with framing
// Reads framed SerializationTestMessage from stdin and decodes it
int main() {
  // Read binary data from stdin
  uint8_t buffer[512];
  size_t size = fread(buffer, 1, sizeof(buffer), stdin);

  if (size == 0) {
    fprintf(stderr, "ERROR: No data received from stdin\n");
    return 1;
  }

  // Decode the message
  packet_format_t* format = &default_frame_format;
  msg_info_t decode_result = format->validate_packet(buffer, size);

  if (!decode_result.valid) {
    fprintf(stderr, "ERROR: Failed to decode message\n");
    return 1;
  }

  SerializationTestSerializationTestMessage msg =
      serialization_test_serialization_test_message_get(decode_result);

  // Print decoded values in a consistent format
  printf("magic_number=0x%08X\n", msg.magic_number);
  printf("test_string=%.*s\n", msg.test_string.length, msg.test_string.data);
  printf("test_float=%.5f\n", msg.test_float);
  printf("test_bool=%s\n", msg.test_bool ? "True" : "False");
  printf("test_array=");
  for (int i = 0; i < msg.test_array.count; i++) {
    printf("%d", msg.test_array.data[i]);
    if (i < msg.test_array.count - 1) printf(",");
  }
  printf("\n");

  return 0;
}

File: tests/ts/decoder_framed.ts
#!/usr/bin/env node
// Decoder for cross-platform testing with framing
// Reads framed SerializationTestMessage from stdin and decodes it

const serializationTestModule = require('./serialization_test.sf');
const structFrameParserModule = require('./struct_frame_parser');

const serialization_test_SerializationTestMessage = serializationTestModule.serialization_test_SerializationTestMessage;
const FrameParser = structFrameParserModule.FrameParser;
const BasicPacketFormat = structFrameParserModule.BasicPacketFormat;

try {
  // Set up stdin to read binary data
  const chunks: Buffer[] = [];
  process.stdin.on('data', (chunk) => {
    chunks.push(chunk);
  });

  process.stdin.on('end', () => {
    const binaryData = Buffer.concat(chunks);

    if (binaryData.length === 0) {
      process.stderr.write('ERROR: No data received from stdin\n');
      process.exit(1);
    }

    // Create parser for deserialization
    const packetFormats: { [key: number]: any } = {
      0x90: new BasicPacketFormat()
    };
    const msgDefinitions: { [key: number]: any } = {
      204: serialization_test_SerializationTestMessage
    };

    const parser = new FrameParser(packetFormats, msgDefinitions);

    // Parse the binary data byte by byte
    let msg: any = null;
    for (const byte of binaryData) {
      const result = parser.parse_char(byte);
      if (result) {
        msg = result;
        break;
      }
    }

    if (msg === null) {
      process.stderr.write('ERROR: Failed to decode message\n');
      process.exit(1);
    }

    // Print decoded values in a consistent format
    console.log(`magic_number=0x${msg.magic_number.toString(16).toUpperCase().padStart(8, '0')}`);
    console.log(`test_string=${msg.test_string_data.substring(0, msg.test_string_length)}`);
    console.log(`test_float=${msg.test_float.toFixed(5)}`);
    console.log(`test_bool=${msg.test_bool}`);
    console.log(`test_array=${msg.test_array_data.slice(0, msg.test_array_count).join(',')}`);
  });

  // Resume stdin to start reading
  process.stdin.resume();
} catch (error) {
  process.stderr.write(`ERROR: Failed to decode: ${error}\n`);
  process.exit(1);
}
